odoo-openmrs/src/main/java/com/ozonehis/eip/odoo/openmrs/client/OdooClient.java

```diff
@@ -0,0 +1,172 @@
+ */
+package com.ozonehis.eip.odoo.openmrs.client;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+
+import com.ozonehis.eip.odoo.openmrs.Constants;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.*;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.xmlrpc.XmlRpcException;
+import org.apache.xmlrpc.client.XmlRpcClient;
+import org.apache.xmlrpc.client.XmlRpcClientConfigImpl;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Getter
+@Component
+public class OdooClient {
+
+    @Value("${odoo.baseUrl}")
+    private String url;
+
+    @Value("${odoo.database}")
+    private String database;
+
+    @Value("${odoo.username}")
+    private String username;
+
+    @Value("${odoo.password}")
+    private String password;
+
+    private Integer uid;
+
+    private XmlRpcClientConfigImpl xmlRpcClientConfig;
+
+    private XmlRpcClient client;
+
+    private static final String SERVER_OBJECT_URL = "%s/xmlrpc/2/object";
+
+    private static final String SERVER_COMMON_URL = "%s/xmlrpc/2/common";
+
+    public OdooClient() {}
+
+    public OdooClient(String url, String database, String username, String password) {
+        this.url = url;
+        this.database = database;
+        this.username = username;
+        this.password = password;
+    }
+
+    public void init() {
+        if (xmlRpcClientConfig == null) {
+            try {
+                xmlRpcClientConfig = new XmlRpcClientConfigImpl();
+                xmlRpcClientConfig.setEnabledForExtensions(true);
+                xmlRpcClientConfig.setServerURL(new URL(String.format(SERVER_OBJECT_URL, getUrl())));
+            } catch (MalformedURLException e) {
+                log.error("Error occurred while building odoo server url {} error {}", getUrl(), e.getMessage());
+                throw new RuntimeException(e);
+            }
+        }
+        if (client == null) {
+            client = new XmlRpcClient();
+            client.setConfig(xmlRpcClientConfig);
+        }
+        // authenticate
+        if (uid == null) {
+            try {
+                XmlRpcClientConfigImpl xmlRpcClientCommonConfig = new XmlRpcClientConfigImpl();
+                xmlRpcClientCommonConfig.setServerURL(new URL(String.format(SERVER_COMMON_URL, getUrl())));
+                uid = (Integer) client.execute(
+                        xmlRpcClientCommonConfig,
+                        "authenticate",
+                        asList(getDatabase(), getUsername(), getPassword(), emptyMap()));
+            } catch (XmlRpcException | MalformedURLException e) {
+                log.error("Cannot authenticate to Odoo server error: {}", e.getMessage());
+                throw new RuntimeException("Cannot authenticate to Odoo server");
```

Comment:
If you're re-throwing the error, you might as well skip logging a message

Reply:
Done